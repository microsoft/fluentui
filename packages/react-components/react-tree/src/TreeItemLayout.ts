export * from './components/TreeItemLayout/index';
