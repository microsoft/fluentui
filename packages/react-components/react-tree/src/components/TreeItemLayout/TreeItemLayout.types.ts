import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { TreeItemContextValue } from '../../contexts/treeItemContext';

export type TreeItemLayoutSlots = {
  root: Slot<'div'>;
  /**
   * Icon slot that renders right before main content
   */
  iconBefore?: Slot<'span'>;
  /**
   * Icon slot that renders right after main content
   */
  iconAfter?: Slot<'span'>;
  aside?: Slot<'span'>;
};

/**
 * TreeItemLayout Props
 */
export type TreeItemLayoutProps = ComponentProps<Partial<TreeItemLayoutSlots>>;

/**
 * State used in rendering TreeItemLayout
 */
export type TreeItemLayoutState = ComponentState<TreeItemLayoutSlots> & TreeItemContextValue;
