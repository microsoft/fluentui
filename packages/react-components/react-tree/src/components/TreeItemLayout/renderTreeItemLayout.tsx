import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TreeItemLayoutState, TreeItemLayoutSlots } from './TreeItemLayout.types';

/**
 * Render the final JSX of TreeItemLayout
 */
export const renderTreeItemLayout_unstable = (state: TreeItemLayoutState) => {
  const { isActionsVisible } = state;
  const { slots, slotProps } = getSlots<TreeItemLayoutSlots>(state);
  return (
    <slots.root {...slotProps.root}>
      {slots.iconBefore && <slots.iconBefore {...slotProps.iconBefore} />}
      {slotProps.root.children}
      {slots.iconAfter && <slots.iconAfter {...slotProps.iconAfter} />}
      {!isActionsVisible && slots.aside && <slots.aside {...slotProps.aside} />}
    </slots.root>
  );
};
