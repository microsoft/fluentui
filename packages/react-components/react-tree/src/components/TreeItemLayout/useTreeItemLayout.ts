import * as React from 'react';
import { getNativeElementProps, resolveShorthand } from '@fluentui/react-utilities';
import type { TreeItemLayoutProps, TreeItemLayoutState } from './TreeItemLayout.types';
import { useTreeItemContext_unstable } from '../../contexts/treeItemContext';

/**
 * Create the state required to render TreeItemLayout.
 *
 * The returned state can be modified with hooks such as useTreeItemLayoutStyles_unstable,
 * before being passed to renderTreeItemLayout_unstable.
 *
 * @param props - props from this instance of TreeItemLayout
 * @param ref - reference to root HTMLElement of TreeItemLayout
 */
export const useTreeItemLayout_unstable = (
  props: TreeItemLayoutProps,
  ref: React.Ref<HTMLElement>,
): TreeItemLayoutState => {
  const { iconAfter, iconBefore, aside, as = 'div' } = props;
  const treeItemContext = useTreeItemContext_unstable();

  return {
    ...treeItemContext,
    components: {
      root: 'div',
      iconBefore: 'span',
      iconAfter: 'span',
      aside: 'span',
    },
    root: getNativeElementProps(as, { ...props, ref }),
    iconBefore: resolveShorthand(iconBefore, { defaultProps: { 'aria-hidden': true } }),
    iconAfter: resolveShorthand(iconAfter, { defaultProps: { 'aria-hidden': true } }),
    aside: resolveShorthand(aside, { defaultProps: { 'aria-hidden': true } }),
  };
};
