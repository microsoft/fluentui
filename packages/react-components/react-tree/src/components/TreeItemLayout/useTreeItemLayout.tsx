import * as React from 'react';
import {
  ExtractSlotProps,
  getIntrinsicElementProps,
  isResolvedShorthand,
  useMergedRefs,
  slot,
} from '@fluentui/react-utilities';
import { useTreeItemContext_unstable, useTreeContext_unstable } from '../../contexts';
import type { TreeItemLayoutProps, TreeItemLayoutSlots, TreeItemLayoutState } from './TreeItemLayout.types';
import { Checkbox, CheckboxProps } from '@fluentui/react-checkbox';
import { Radio, RadioProps } from '@fluentui/react-radio';
import { TreeItemChevron } from '../TreeItemChevron';
import { useArrowNavigationGroup } from '@fluentui/react-tabster';

/**
 * Create the state required to render TreeItemLayout.
 *
 * The returned state can be modified with hooks such as useTreeItemLayoutStyles_unstable,
 * before being passed to renderTreeItemLayout_unstable.
 *
 * @param props - props from this instance of TreeItemLayout
 * @param ref - reference to root HTMLElement of TreeItemLayout
 */
export const useTreeItemLayout_unstable = (
  props: TreeItemLayoutProps,
  ref: React.Ref<HTMLElement>,
): TreeItemLayoutState => {
  const { main, iconAfter, iconBefore } = props;

  const layoutRef = useTreeItemContext_unstable(ctx => ctx.layoutRef);
  const selectionMode = useTreeContext_unstable(ctx => ctx.selectionMode);

  const [isActionsVisibleExternal, actionsShorthand]: [boolean | undefined, TreeItemLayoutSlots['actions']] =
    isResolvedShorthand(props.actions)
      ? // .visible prop should not be propagated to the DOM
        [props.actions.visible, { ...props.actions, visible: undefined }]
      : [undefined, props.actions];

  const isActionsVisible = useTreeItemContext_unstable(ctx => ctx.isActionsVisible) || isActionsVisibleExternal;
  const isAsideVisible = useTreeItemContext_unstable(ctx => ctx.isAsideVisible);
  const selectionRef = useTreeItemContext_unstable(ctx => ctx.selectionRef);
  const expandIconRef = useTreeItemContext_unstable(ctx => ctx.expandIconRef);
  const actionsRef = useTreeItemContext_unstable(ctx => ctx.actionsRef);
  const checked = useTreeItemContext_unstable(ctx => ctx.checked);
  const isBranch = useTreeItemContext_unstable(ctx => ctx.itemType === 'branch');

  const expandIcon = slot.optional(props.expandIcon, {
    renderByDefault: isBranch,
    defaultProps: {
      children: <TreeItemChevron />,
      'aria-hidden': true,
    },
    elementType: 'div',
  });
  const expandIconRefs = useMergedRefs(expandIcon?.ref, expandIconRef);
  if (expandIcon) {
    expandIcon.ref = expandIconRefs;
  }
  const arrowNavigationProps = useArrowNavigationGroup({ circular: true, axis: 'horizontal' });
  const actions = isActionsVisible
    ? slot.optional(actionsShorthand, {
        defaultProps: { ...arrowNavigationProps, role: 'toolbar' } as ExtractSlotProps<TreeItemLayoutSlots['actions']>,
        elementType: 'div',
      })
    : undefined;
  const actionsRefs = useMergedRefs(actions?.ref, actionsRef);
  if (actions) {
    actions.ref = actionsRefs;
  }
  return {
    components: {
      root: 'div',
      expandIcon: 'div',
      iconBefore: 'div',
      main: 'div',
      iconAfter: 'div',
      actions: 'div',
      aside: 'div',
      // Casting here to a union between checkbox and radio
      selector: (selectionMode === 'multiselect' ? Checkbox : Radio) as React.ElementType<CheckboxProps | RadioProps>,
    },
    buttonContextValue: { size: 'small' },
    root: slot.always(
      getIntrinsicElementProps('div', {
        ...props,
        // FIXME:
        // `ref` is wrongly assigned to be `HTMLElement` instead of `HTMLDivElement`
        // but since it would be a breaking change to fix it, we are casting ref to it's proper type
        ref: useMergedRefs(ref, layoutRef) as React.Ref<HTMLDivElement>,
      }),
      {
        elementType: 'div',
      },
    ),
    iconBefore: slot.optional(iconBefore, { defaultProps: { 'aria-hidden': true }, elementType: 'div' }),
    main: slot.always(main, { elementType: 'div' }),
    iconAfter: slot.optional(iconAfter, { defaultProps: { 'aria-hidden': true }, elementType: 'div' }),
    aside: isAsideVisible
      ? slot.optional(props.aside, { defaultProps: { 'aria-hidden': true }, elementType: 'div' })
      : undefined,
    actions,
    expandIcon,
    selector: slot.optional(props.selector, {
      renderByDefault: selectionMode !== 'none',
      defaultProps: {
        checked,
        tabIndex: -1,
        'aria-hidden': true,
        ref: selectionRef,
        // casting here to a union between checkbox and radio
        // since ref is not present on the selector signature
        // FIXME: look into Slot type to see if we can make this work
      } as CheckboxProps | RadioProps,
      elementType: (selectionMode === 'multiselect' ? Checkbox : Radio) as React.ElementType<
        CheckboxProps | RadioProps
      >,
    }),
  };
};
