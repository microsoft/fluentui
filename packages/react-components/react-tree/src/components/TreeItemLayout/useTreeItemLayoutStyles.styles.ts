import { makeResetStyles, makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TreeItemLayoutSlots, TreeItemLayoutState } from './TreeItemLayout.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens, typographyStyles } from '@fluentui/react-theme';
import { useTreeContext_unstable } from '../../contexts/treeContext';
import { treeItemLevelToken } from '../../utils/tokens';
import { useTreeItemContext_unstable } from '../../contexts/treeItemContext';

export const treeItemLayoutClassNames: SlotClassNames<TreeItemLayoutSlots> = {
  root: 'fui-TreeItemLayout',
  iconBefore: 'fui-TreeItemLayout__iconBefore',
  main: 'fui-TreeItemLayout__main',
  iconAfter: 'fui-TreeItemLayout__iconAfter',
  expandIcon: 'fui-TreeItemLayout__expandIcon',
  aside: 'fui-TreeItemLayout__aside',
  actions: 'fui-TreeItemLayout__actions',
  selector: 'fui-TreeItemLayout__selector',
};

const useRootBaseStyles = makeResetStyles({
  display: 'flex',
  alignItems: 'center',
  minHeight: '32px',
  boxSizing: 'border-box',
  ...shorthands.gridArea('layout'),
  ':active': {
    color: tokens.colorNeutralForeground2Pressed,
    backgroundColor: tokens.colorSubtleBackgroundPressed,
    // TODO: stop using treeItemLayoutClassNames.expandIcon for styling
    [`& .${treeItemLayoutClassNames.expandIcon}`]: {
      color: tokens.colorNeutralForeground3Pressed,
    },
  },
  ':hover': {
    color: tokens.colorNeutralForeground2Hover,
    backgroundColor: tokens.colorSubtleBackgroundHover,
    // TODO: stop using treeItemLayoutClassNames.expandIcon  for styling
    [`& .${treeItemLayoutClassNames.expandIcon}`]: {
      color: tokens.colorNeutralForeground3Hover,
    },
  },
});

/**
 * Styles for the root slot
 */
const useRootStyles = makeStyles({
  leaf: {
    paddingLeft: `calc(var(${treeItemLevelToken}, 1) * ${tokens.spacingHorizontalXXL})`,
  },
  branch: {
    paddingLeft: `calc((var(${treeItemLevelToken}, 1) - 1) * ${tokens.spacingHorizontalXXL})`,
  },
  medium: {
    ...typographyStyles.body1,
  },
  small: {
    minHeight: '24px',
    ...typographyStyles.caption1,
  },
  // Appearance variations
  subtle: {},
  'subtle-alpha': {
    ':hover': {
      backgroundColor: tokens.colorSubtleBackgroundLightAlphaHover,
    },
    ':active': {
      backgroundColor: tokens.colorSubtleBackgroundLightAlphaPressed,
    },
  },
  transparent: {
    backgroundColor: tokens.colorTransparentBackground,
    ':hover': {
      backgroundColor: tokens.colorTransparentBackgroundHover,
    },
    ':active': {
      backgroundColor: tokens.colorTransparentBackgroundPressed,
    },
  },
});

/**
 * Styles for the action icon slot
 */
const useActionsBaseStyles = makeResetStyles({
  display: 'flex',
  marginLeft: 'auto',
  position: 'relative',
  zIndex: 1,
  ...shorthands.gridArea('aside'),
  ...shorthands.padding(0, tokens.spacingHorizontalS),
});
/**
 * Styles for the action icon slot
 */
const useAsideBaseStyles = makeResetStyles({
  display: 'flex',
  marginLeft: 'auto',
  alignItems: 'center',
  zIndex: 0,
  ...shorthands.gridArea('aside'),
  ...shorthands.padding(0, tokens.spacingHorizontalM),
  ...shorthands.gap(tokens.spacingHorizontalXS),
});

/**
 * Styles for the expand icon slot
 */
const useExpandIconBaseStyles = makeResetStyles({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  minWidth: '24px',
  boxSizing: 'border-box',
  color: tokens.colorNeutralForeground3,
  ...shorthands.flex(0, 0, 'auto'),
  ...shorthands.padding(tokens.spacingVerticalXS, 0),
});

/**
 * Styles for the content slot
 */
const useMainBaseStyles = makeResetStyles({
  ...shorthands.padding(0, tokens.spacingHorizontalXXS),
});

/**
 * Styles for the before/after icon slot
 */
const useIconBaseStyles = makeResetStyles({
  display: 'flex',
  alignItems: 'center',
  color: tokens.colorNeutralForeground2,
  lineHeight: tokens.lineHeightBase500,
  fontSize: tokens.fontSizeBase500,
});

const useIconBeforeStyles = makeStyles({
  medium: {
    paddingRight: tokens.spacingHorizontalXS,
  },
  small: {
    paddingRight: tokens.spacingHorizontalXXS,
  },
});

const useIconAfterStyles = makeStyles({
  medium: {
    paddingLeft: tokens.spacingHorizontalXS,
  },
  small: {
    paddingLeft: tokens.spacingHorizontalXXS,
  },
});

/**
 * Apply styling to the TreeItemLayout slots based on the state
 */
export const useTreeItemLayoutStyles_unstable = (state: TreeItemLayoutState): TreeItemLayoutState => {
  const { main, iconAfter, iconBefore, expandIcon, root, aside, actions, selector } = state;
  const rootStyles = useRootStyles();
  const rootBaseStyles = useRootBaseStyles();
  const actionsBaseStyles = useActionsBaseStyles();
  const asideBaseStyles = useAsideBaseStyles();

  const mainBaseStyles = useMainBaseStyles();

  const expandIconBaseStyles = useExpandIconBaseStyles();
  const iconBaseStyles = useIconBaseStyles();
  const iconBeforeStyles = useIconBeforeStyles();
  const iconAfterStyles = useIconAfterStyles();

  const size = useTreeContext_unstable(ctx => ctx.size);
  const appearance = useTreeContext_unstable(ctx => ctx.appearance);
  const itemType = useTreeItemContext_unstable(ctx => ctx.itemType);

  root.className = mergeClasses(
    treeItemLayoutClassNames.root,
    rootBaseStyles,
    rootStyles[appearance],
    rootStyles[size],
    rootStyles[itemType],
    root.className,
  );

  main.className = mergeClasses(treeItemLayoutClassNames.main, mainBaseStyles, main.className);

  if (expandIcon) {
    expandIcon.className = mergeClasses(
      treeItemLayoutClassNames.expandIcon,
      expandIconBaseStyles,
      expandIcon.className,
    );
  }

  if (iconBefore) {
    iconBefore.className = mergeClasses(
      treeItemLayoutClassNames.iconBefore,
      iconBaseStyles,
      iconBeforeStyles[size],
      iconBefore.className,
    );
  }

  if (iconAfter) {
    iconAfter.className = mergeClasses(
      treeItemLayoutClassNames.iconAfter,
      iconBaseStyles,
      iconAfterStyles[size],
      iconAfter.className,
    );
  }

  if (actions) {
    actions.className = mergeClasses(treeItemLayoutClassNames.actions, actionsBaseStyles, actions.className);
  }
  if (aside) {
    aside.className = mergeClasses(treeItemLayoutClassNames.aside, asideBaseStyles, aside.className);
  }
  if (selector) {
    selector.className = mergeClasses(treeItemLayoutClassNames.selector, selector.className);
  }

  return state;
};
