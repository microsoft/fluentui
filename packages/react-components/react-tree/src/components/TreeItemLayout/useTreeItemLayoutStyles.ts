import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TreeItemLayoutSlots, TreeItemLayoutState } from './TreeItemLayout.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens, typographyStyles } from '@fluentui/react-theme';
import { useTreeContext_unstable } from '../../contexts/treeContext';

export const treeItemLayoutClassNames: SlotClassNames<TreeItemLayoutSlots> = {
  root: 'fui-TreeItemLayout',
  aside: 'fui-TreeItemLayout__aside',
  iconAfter: 'fui-TreeItemLayout__iconAfter',
  iconBefore: 'fui-TreeItemLayout__iconBefore',
};

/**
 * Styles for the root slot
 */
const useRootStyles = makeStyles({
  base: {
    display: 'flex',
    alignItems: 'center',
    minHeight: '32px',
    ...shorthands.flex(1),
  },
  medium: {
    columnGap: tokens.spacingHorizontalSNudge,
    ...typographyStyles.body1,
  },
  small: {
    columnGap: tokens.spacingHorizontalXS,
    minHeight: '24px',
    ...typographyStyles.caption1,
  },
});

/**
 * Styles for the before/after icon slot
 */
const useIconStyles = makeStyles({
  base: {
    display: 'flex',
    alignItems: 'center',
    color: tokens.colorNeutralForeground2,
    lineHeight: tokens.lineHeightBase500,
    fontSize: tokens.fontSizeBase500,
  },
});

/**
 * Styles for the action icon slot
 */
const useAsideStyles = makeStyles({
  base: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: 'auto',
    ...shorthands.padding(0, tokens.spacingHorizontalM),
    ...shorthands.gap(tokens.spacingHorizontalXS),
  },
});

/**
 * Apply styling to the TreeItemLayout slots based on the state
 */
export const useTreeItemLayoutStyles_unstable = (state: TreeItemLayoutState): TreeItemLayoutState => {
  const { iconAfter, iconBefore, aside, root } = state;
  const rootStyles = useRootStyles();
  const iconStyles = useIconStyles();
  const asideStyles = useAsideStyles();

  const size = useTreeContext_unstable(ctx => ctx.size);

  root.className = mergeClasses(treeItemLayoutClassNames.root, rootStyles.base, rootStyles[size], root.className);

  if (iconBefore) {
    iconBefore.className = mergeClasses(treeItemLayoutClassNames.iconBefore, iconStyles.base, iconBefore.className);
  }

  if (iconAfter) {
    iconAfter.className = mergeClasses(treeItemLayoutClassNames.iconAfter, iconStyles.base, iconAfter.className);
  }

  if (aside) {
    aside.className = mergeClasses(treeItemLayoutClassNames.aside, asideStyles.base, aside.className);
  }

  return state;
};
