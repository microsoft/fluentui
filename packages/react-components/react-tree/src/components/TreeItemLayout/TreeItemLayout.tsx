import * as React from 'react';
import { useTreeItemLayout_unstable } from './useTreeItemLayout';
import { renderTreeItemLayout_unstable } from './renderTreeItemLayout';
import { useTreeItemLayoutStyles_unstable } from './useTreeItemLayoutStyles';
import type { TreeItemLayoutProps } from './TreeItemLayout.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * TreeItemLayout component - TODO: add more docs
 */
export const TreeItemLayout: ForwardRefComponent<TreeItemLayoutProps> = React.forwardRef((props, ref) => {
  const state = useTreeItemLayout_unstable(props, ref);

  useTreeItemLayoutStyles_unstable(state);
  return renderTreeItemLayout_unstable(state);
});

TreeItemLayout.displayName = 'TreeItemLayout';
