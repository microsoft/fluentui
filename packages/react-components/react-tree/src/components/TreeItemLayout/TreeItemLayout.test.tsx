import * as React from 'react';
import { render } from '@testing-library/react';
import { TreeItemLayout } from './TreeItemLayout';
import { isConformant } from '../../testing/isConformant';

describe('TreeItemLayout', () => {
  isConformant({
    Component: TreeItemLayout,
    displayName: 'TreeItemLayout',
    requiredProps: {
      aside: 'aside',
      iconAfter: 'iconAfter',
      iconBefore: 'iconBefore',
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TreeItemLayout>Default TreeItemLayout</TreeItemLayout>);
    expect(result.container).toMatchSnapshot();
  });
});
