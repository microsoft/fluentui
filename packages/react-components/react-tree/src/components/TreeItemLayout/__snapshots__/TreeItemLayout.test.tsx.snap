// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeItemLayout renders a default state 1`] = `
<div>
  <div
    class="fui-TreeItemLayout"
  >
    Default TreeItemLayout
  </div>
</div>
`;
